DataFormats/TrackerRecHit2D/interface/VectorHit.h

```diff
@@ -0,0 +1,160 @@
+            OmniClusterRef const& upper);
+
+  ~VectorHit() override;
+
+  VectorHit* clone() const override { return new VectorHit(*this); }
+#ifndef __GCCXML__
+  RecHitPointer cloneSH() const override { return std::make_shared<VectorHit>(*this); }
+#endif
+
+  bool sharesInput(const TrackingRecHit* other, SharedInputType what) const override;
+  bool sharesClusters(VectorHit const& h1, VectorHit const& h2, SharedInputType what) const;
+
+  // Parameters of the segment, for the track fit
+  // For a 4D segment: (dx/dz,dy/dz,x,y)
+  bool hasPositionAndError() const override {
+    //bool hasPositionAndError() const {
+    return true;
+    //      return (err_.xx() != 0) || (err_.yy() != 0) || (err_.xy() != 0) ||
+    //             (pos_.x()  != 0) || (pos_.y()  != 0) || (pos_.z()  != 0);
+  };
+
+  AlgebraicVector parameters() const override;
+  void getKfComponents(KfComponentsHolder& holder) const override { getKfComponents4D(holder); }
+  void getKfComponents4D(KfComponentsHolder& holder) const;
+
+  // returning methods
+  LocalPoint localPosition() const override { return thePosition; }
+  virtual LocalVector localDirection() const { return theDirection; }
+  AlgebraicSymMatrix parametersError() const override;
+  LocalError localPositionError() const override;
+  virtual LocalError localDirectionError() const;
+  Global3DVector globalDirection() const;
+
+  virtual double chi2() const { return theChi2; }
+  int dimension() const override { return theDimension; }
+
+  std::pair<double, double> curvatureORphi(std::string curvORphi = "curvature") const;
+
+  ClusterRef lowerCluster() const { return theLowerCluster.cluster_phase2OT(); }
+  ClusterRef upperCluster() const { return theUpperCluster.cluster_phase2OT(); }
+  OmniClusterRef const lowerClusterRef() const { return theLowerCluster; }
+  OmniClusterRef const upperClusterRef() const { return theUpperCluster; }
+
+  //FIXME::to update with a proper CPE maybe...
+  Global3DPoint lowerGlobalPos() const;
+  Global3DPoint upperGlobalPos() const;
+  Global3DPoint phase2clusterGlobalPos(const PixelGeomDetUnit* geomDet, ClusterRef cluster) const;
+  GlobalError lowerGlobalPosErr() const;
+  GlobalError upperGlobalPosErr() const;
+  GlobalError phase2clusterGlobalPosErr(const PixelGeomDetUnit* geomDet) const;
+
+  bool isPhase2() const override { return true; }
+
+  //FIXME: I have always two clusters in a VH
+  OmniClusterRef const& firstClusterRef() const override { return theLowerCluster; }
+  ClusterRef cluster() const { return theLowerCluster.cluster_phase2OT(); }
+
+  //This method returns the delta in global coordinates
+  Global3DVector globalDelta() const;
+  float theta();
+
+  /// The projection matrix relates the trajectory state parameters to the segment parameters().
+  AlgebraicMatrix projectionMatrix() const override;
+
+  // Access to component RecHits (if any)
+  std::vector<const TrackingRecHit*> recHits() const override;
+  std::vector<TrackingRecHit*> recHits() override;
```

Comment:
what is this non const method that return something by value????

Reply:
Not having it leads to compilation errors because the concrete implementation of the non-const function seems to needed, without it I get the following errors:

`/uscms_data/d3/jschulte/VH/new/CMSSW_11_2_X_2020-09-03-1100/src/DataFormats/TrackerRecHit2D/interface/VectorHit.h:52:65: error: invalid new-expression of abstract class type 'VectorHit'
   VectorHit* clone() const override { return new VectorHit(*this); }
                                                                 ^
/uscms_data/d3/jschulte/VH/new/CMSSW_11_2_X_2020-09-03-1100/src/DataFormats/TrackerRecHit2D/interface/VectorHit.h:28:7: note:   because the following virtual functions are pure within 'VectorHit':
 class VectorHit final : public BaseTrackerRecHit {
       ^~~~~~~~~
In file included from /uscms_data/d3/jschulte/VH/new/CMSSW_11_2_X_2020-09-03-1100/src/DataFormats/TrackerRecHit2D/interface/trackerHitRTTI.h:4,
                 from /uscms_data/d3/jschulte/VH/new/CMSSW_11_2_X_2020-09-03-1100/src/DataFormats/TrackerRecHit2D/interface/BaseTrackerRecHit.h:4,
                 from /uscms_data/d3/jschulte/VH/new/CMSSW_11_2_X_2020-09-03-1100/src/DataFormats/TrackerRecHit2D/interface/VectorHit.h:13,
                 from /uscms_data/d3/jschulte/VH/new/CMSSW_11_2_X_2020-09-03-1100/src/DataFormats/TrackerRecHit2D/src/VectorHit.cc:1:
/cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc820/cms/cmssw-patch/CMSSW_11_2_X_2020-09-03-1100/src/DataFormats/TrackingRecHit/interface/TrackingRecHit.h:105:40: note:      'virtual std::vector<TrackingRecHit*> TrackingRecHit::recHits()'
   virtual std::vector<TrackingRecHit*> recHits() = 0;`